contracts/FlexibleStorage.sol

```
@@ -0,0 +1,166 @@
+pragma solidity ^0.5.16;
+
+// Internal References
+import "./interfaces/IAddressResolver.sol";
+
+
+// https://docs.synthetix.io/contracts/source/contracts/FlexibleStorage
+contract FlexibleStorage {
+    IAddressResolver public resolverProxy;
+
+    mapping(bytes32 => bytes32) public hashes;
+
+    mapping(bytes32 => mapping(bytes32 => uint)) internal UIntStorage;
+    mapping(bytes32 => mapping(bytes32 => address)) internal AddressStorage;
+
+    // mapping(bytes32 => string) internal StringStorage;
+    // mapping(bytes32 => bytes) internal BytesStorage;
+    // mapping(bytes32 => bytes32) internal Bytes32Storage;
+    // mapping(bytes32 => bool) internal BooleanStorage;
+    // mapping(bytes32 => int) internal IntStorage;
+
+    constructor(address _resolver) public {
+        // ReadProxyAddressResolver
+        resolverProxy = IAddressResolver(_resolver);
+    }
+
+    /* ========== INTERNAL FUNCTIONS ========== */
+
+    function _memoizeHash(bytes32 contractName) internal returns (bytes32) {
+        if (hashes[contractName] == bytes32(0)) {
+            // set to unique hash at the time of creation
+            hashes[contractName] = keccak256(abi.encodePacked(msg.sender, contractName, block.number));
+        }
+        return hashes[contractName];
+    }
+
+    function _setUIntValue(
+        bytes32 contractName,
+        bytes32 record,
+        uint value
+    ) internal {
+        UIntStorage[_memoizeHash(contractName)][record] = value;
+        emit ValueSetUInt(contractName, record, value);
+    }
+
+    function _setAddressValue(
+        bytes32 contractName,
+        bytes32 record,
+        address value
+    ) internal {
+        AddressStorage[_memoizeHash(contractName)][record] = value;
+        emit ValueSetAddress(contractName, record, value);
+    }
+
+    /* ========== VIEWS ========== */
+
+    function getUIntValue(bytes32 contractName, bytes32 record) external view returns (uint) {
```

Might be safe to revert if hashes[contractName] = 0x0 to avoid silent failures if a requested value does not exist, instead of returning 0.

---

To implement something like this I'd recommend something like:
```
    function getUIntValue(bytes32 contractName, bytes32 record) external view returns (uint) {
        return UIntStorage[_getHash(contractName)][record];
    }

    function _getHash(bytes32 contractName) internal view returns(bytes32 hash) {
		hash = hashes[contractName];
		require(hash != bytes32(0), "No hash found for the required contractName.");
    }
```

---

Continuing discussion in #622